# Assignment 05 - Population
By: _Daniel Finnerty_



The purpose of this notebook is to analyse the differences between the sexes by age in Ireland. Data is gathered from the below CSO location:

https://ws.cso.ie/public/api.restful/PxStat.Data.Cube_API.ReadDataset/FY006A/CSV/1.0/en

## Packages

Import required package for effective operation of notebook.

In [410]:
# Import Pandas for data frames
import pandas as pd

# Import Numpy
import numpy as np

## Part 1

Download data file as CSV

In [411]:
# Data location
url = "https://ws.cso.ie/public/api.restful/PxStat.Data.Cube_API.ReadDataset/FY006A/CSV/1.0/en"

# Read CVS into a data frame
df = pd.read_csv(url)

# Show first 3 rows
df.head(3)

STATISTIC Statistic Label  TLIST(A1)  CensusYear C02199V02655         Sex  \
0  FY006AC01      Population       2022        2022            -  Both sexes   
1  FY006AC01      Population       2022        2022            -  Both sexes   
2  FY006AC01      Population       2022        2022            -  Both sexes   

  C02076V03371 Single Year of Age                          C03789V04537  \
0            -           All ages                                   IE0   
1            -           All ages  2ae19629-1492-13a3-e055-000000000001   
2            -           All ages  2ae19629-1433-13a3-e055-000000000001   

  Administrative Counties    UNIT    VALUE  
0                 Ireland  Number  5149139  
1   Carlow County Council  Number    61968  
2     Dublin City Council  Number   592713

In [412]:
# Show last 3 rows
df.tail(3)

STATISTIC Statistic Label  TLIST(A1)  CensusYear C02199V02655     Sex  \
9789  FY006AC01      Population       2022        2022            2  Female   
9790  FY006AC01      Population       2022        2022            2  Female   
9791  FY006AC01      Population       2022        2022            2  Female   

     C02076V03371  Single Year of Age                          C03789V04537  \
9789          650  100 years and over  2ae19629-149d-13a3-e055-000000000001   
9790          650  100 years and over  2ae19629-14a4-13a3-e055-000000000001   
9791          650  100 years and over  2ae19629-1495-13a3-e055-000000000001   

      Administrative Counties    UNIT  VALUE  
9789     Cavan County Council  Number     12  
9790   Donegal County Council  Number     31  
9791  Monaghan County Council  Number      7

To determine the weighted mean age by sex, the 'Both sexes' information can be removed.

In [413]:
# Show only data that does not have 'Both sexes' in the Sex column
df = df[df["Sex"] != "Both sexes"]

# And data that does not have 'All ages' in the Single Year of Age column
df = df[df["Single Year of Age"] != "All ages"]

# Remove counties breakdown, and focus solely on the entire country
df = df[df["Administrative Counties"] == "Ireland"]

# Show first 3 rows to confirm
df.head(3)

STATISTIC Statistic Label  TLIST(A1)  CensusYear C02199V02655   Sex  \
3296  FY006AC01      Population       2022        2022            1  Male   
3328  FY006AC01      Population       2022        2022            1  Male   
3360  FY006AC01      Population       2022        2022            1  Male   

     C02076V03371 Single Year of Age C03789V04537 Administrative Counties  \
3296          200       Under 1 year          IE0                 Ireland   
3328          001             1 year          IE0                 Ireland   
3360          002            2 years          IE0                 Ireland   

        UNIT  VALUE  
3296  Number  29610  
3328  Number  28875  
3360  Number  30236

It can be seen that the data has been cleaned, but the age now also needs to be converted solely to integers.

In [414]:
# Change 'Under 1 year' to '0'
df["Single Year of Age"] = df["Single Year of Age"].str.replace('Under 1 year', '0')

# Remove all non-digit characters, leaving only the number
df["Single Year of Age"] = df["Single Year of Age"].str.replace('\D', '', regex=True)

# Convert all age values to 64-bit integers
df['Single Year of Age']=df['Single Year of Age'].astype('int64')

# Convert all 'Value' numbers to 64-bit integers
df['VALUE']=df['VALUE'].astype('int64')

df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 202 entries, 3296 to 9760
Data columns (total 12 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   STATISTIC                202 non-null    object
 1   Statistic Label          202 non-null    object
 2   TLIST(A1)                202 non-null    int64 
 3   CensusYear               202 non-null    int64 
 4   C02199V02655             202 non-null    object
 5   Sex                      202 non-null    object
 6   C02076V03371             202 non-null    object
 7   Single Year of Age       202 non-null    int64 
 8   C03789V04537             202 non-null    object
 9   Administrative Counties  202 non-null    object
 10  UNIT                     202 non-null    object
 11  VALUE                    202 non-null    int64 
dtypes: int64(4), object(8)
memory usage: 20.5+ KB


<>:5: SyntaxWarning: invalid escape sequence '\D'
<>:5: SyntaxWarning: invalid escape sequence '\D'
C:\Users\danie\AppData\Local\Temp\ipykernel_19848\2827833747.py:5: SyntaxWarning: invalid escape sequence '\D'
  df["Single Year of Age"] = df["Single Year of Age"].str.replace('\D', '', regex=True)


With this done, the data can now be simplified, by removing unnecessary columns

In [415]:
# Comfirm all column names
headers = df.columns.tolist()

# Show list
headers

['STATISTIC',
 'Statistic Label',
 'TLIST(A1)',
 'CensusYear',
 'C02199V02655',
 'Sex',
 'C02076V03371',
 'Single Year of Age',
 'C03789V04537',
 'Administrative Counties',
 'UNIT',
 'VALUE']

In [416]:
# Remove unrequired columns
drop_col_list = ['STATISTIC', 'Statistic Label','TLIST(A1)','CensusYear','C02199V02655','C02076V03371','C03789V04537','UNIT']
df.drop(columns=drop_col_list, inplace=True)

# Show remaining details
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 202 entries, 3296 to 9760
Data columns (total 4 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   Sex                      202 non-null    object
 1   Single Year of Age       202 non-null    int64 
 2   Administrative Counties  202 non-null    object
 3   VALUE                    202 non-null    int64 
dtypes: int64(2), object(2)
memory usage: 7.9+ KB


In [417]:
# Data can now be converted to pivot table
df_anal = pd.pivot_table(df, 'VALUE',"Single Year of Age","Sex")

#Show
print (df_anal.head(3))

# write out the entire file to local machine
#df_anal.to_csv("population_for_analysis.csv")

Sex                  Female     Male
Single Year of Age                  
0                   28186.0  29610.0
1                   27545.0  28875.0
2                   28974.0  30236.0


In [418]:
# Show first 5 rows
df.head()

Sex  Single Year of Age Administrative Counties  VALUE
3296  Male                   0                 Ireland  29610
3328  Male                   1                 Ireland  28875
3360  Male                   2                 Ireland  30236
3392  Male                   3                 Ireland  31001
3424  Male                   4                 Ireland  31686

This can now be converted to a pivot table, clearly showing the number of each gender for each age

In [419]:
# Create pivot table of df data frame.
df_anal = pd.pivot_table(df, 'VALUE',"Single Year of Age","Sex")

# Show first 3 rows
print (df_anal.head(3))

Sex                  Female     Male
Single Year of Age                  
0                   28186.0  29610.0
1                   27545.0  28875.0
2                   28974.0  30236.0


Next step is to generate seperate groupings by gender.

In [420]:
# Confirm column names
genders = list(df_anal.columns)

# Show
genders[0:]


['Female', 'Male']

In [421]:
# Specifty female column
female_col = genders[0]

# Confirm it is the female column
female_col

'Female'

In [422]:
# Specifty male column
male_col = genders[1]

# Confirm it is the male column
male_col

'Male'

In [423]:
# Calculate weighted mean female age using numpy
female_w_mean = np.average(df_anal.index, weights=df_anal[female_col])

# Show
female_w_mean

38.9397958987787

In [424]:
# Calculate weighted mean male age using numpy
male_w_mean = np.average(df_anal.index, weights=df_anal[male_col])

# Show
male_w_mean

37.7394477371039

It can be seen through this calculation that the weighted mean age for females is 38.94, whereas the same for males is 37.74

To determine the difference between the sexes by age, a column can be added to show the difference in values.

In [425]:
# Add a 'count_diff' column, with the result of subtracting male counts from female
df_anal['count_diff'] = df_anal['Female'] - df_anal['Male']

# Show first 3 rows
df_anal.head(3)

Sex                  Female     Male  count_diff
Single Year of Age                              
0                   28186.0  29610.0     -1424.0
1                   27545.0  28875.0     -1330.0
2                   28974.0  30236.0     -1262.0

From this, any values less than zero, show the quantity of males in that age group exceeds that of females, whereas values greater than zero confirm the opposite.

In [426]:
# Confirm ages where the difference value is less than 0 (more males than females)
df_male = df_anal[df_anal["count_diff"] < 0]

# Show
df_male

Sex                  Female     Male  count_diff
Single Year of Age                              
0                   28186.0  29610.0     -1424.0
1                   27545.0  28875.0     -1330.0
2                   28974.0  30236.0     -1262.0
3                   29483.0  31001.0     -1518.0
4                   29819.0  31686.0     -1867.0
5                   31342.0  32684.0     -1342.0
6                   32622.0  34092.0     -1470.0
7                   33572.0  35429.0     -1857.0
8                   34437.0  36296.0     -1859.0
9                   35227.0  36969.0     -1742.0
10                  36066.0  38035.0     -1969.0
11                  36863.0  38296.0     -1433.0
12                  37297.0  38957.0     -1660.0
13                  36782.0  38474.0     -1692.0
14                  36080.0  37352.0     -1272.0
15                  34492.0  35948.0     -1456.0
16                  33168.0  34646.0     -1478.0
17                  33045.0  34981.0     -1936.0
18                  32852.0  33900.0     -1048.0
19                  31729.0  32867.0     -1138.0
20                  31624.0  32120.0      -496.0
21                  29890.0  30771.0      -881.0
22                  30768.0  31450.0      -682.0
23                  30051.0  30871.0      -820.0
24                  29364.0  30234.0      -870.0
25                  29294.0  29969.0      -675.0
26                  28864.0  29376.0      -512.0
53                  32204.0  32611.0      -407.0

In [427]:
# Confirm ages where the difference value is greated than 0 (more females than males)
df_female = df_anal[df_anal["count_diff"] > 0]

# Show
df_female

Sex                  Female     Male  count_diff
Single Year of Age                              
27                  29528.0  29162.0       366.0
28                  29788.0  28902.0       886.0
29                  30936.0  29989.0       947.0
30                  32841.0  30858.0      1983.0
31                  33710.0  32237.0      1473.0
...                     ...      ...         ...
96                    956.0    327.0       629.0
97                    732.0    217.0       515.0
98                    492.0    130.0       362.0
99                    336.0    105.0       231.0
100                   584.0    154.0       430.0

[73 rows x 3 columns]

It can be seen that from ages zero to 26, and age of 53, there are more males than females however, females outnumber males in all other age groups. One last check can be done to ensure the are no age ranges with the exact same number of males and females (difference equal to zero)

In [428]:
# Confirm ages where the difference value is equal to 0 (same number of females and males)
df_equal = df_anal[df_anal["count_diff"] == 0]

# Show
df_equal

Empty DataFrame
Columns: [Female, Male, count_diff]
Index: []

This shows there are no ages which has the same number of males and females.

## Part 2

Firstly, the data needs to be isolated for the requested age range

In [429]:
# Specify age
age = 35

# specify range in number of years
range = 5

# Create group within above range
df_test = df_anal[(df_anal.index >= (age-range)) & (df_anal.index <= (age+range))]

# Show
df_test

Sex                  Female     Male  count_diff
Single Year of Age                              
30                  32841.0  30858.0      1983.0
31                  33710.0  32237.0      1473.0
32                  34382.0  32413.0      1969.0
33                  34489.0  31888.0      2601.0
34                  36284.0  33121.0      3163.0
35                  37940.0  34695.0      3245.0
36                  39030.0  35828.0      3202.0
37                  39193.0  36427.0      2766.0
38                  40902.0  37513.0      3389.0
39                  42592.0  38749.0      3843.0
40                  43143.0  40301.0      2842.0

The 'count_diff' column can now be used to determine the population difference for this range. As with above, a positive count shows more females than males, where a negative count is the opposite.

In [430]:
# Sum total value of 'count_diff' column
# Source: https://stackoverflow.com/questions/73363606/how-to-sum-the-rows-of-a-dataframe-and-create-a-sub-row-below
diff_total = df_test.iloc[:,2].sum()

# Show
diff_total

30476.0

The same can be done to determine the total population within the grouped range.

In [431]:
# Sum female column (0) and male column (1) together
range_pop_total = (df_test.iloc[:,0].sum())+(df_test.iloc[:,1].sum())

# Show
range_pop_total

798536.0

This shows that within this age range grouping, there are 30,476 more females than males, from a combined population of 798,536 people.

## Part 3

Pull the data again and repeat similar steps to part 1, this time focusing on the county breakdown

In [432]:
# Data location
url = "https://ws.cso.ie/public/api.restful/PxStat.Data.Cube_API.ReadDataset/FY006A/CSV/1.0/en"

# Read CVS into a data frame
df3 = pd.read_csv(url)

# Show first 3 rows
df3.head(3)

STATISTIC Statistic Label  TLIST(A1)  CensusYear C02199V02655         Sex  \
0  FY006AC01      Population       2022        2022            -  Both sexes   
1  FY006AC01      Population       2022        2022            -  Both sexes   
2  FY006AC01      Population       2022        2022            -  Both sexes   

  C02076V03371 Single Year of Age                          C03789V04537  \
0            -           All ages                                   IE0   
1            -           All ages  2ae19629-1492-13a3-e055-000000000001   
2            -           All ages  2ae19629-1433-13a3-e055-000000000001   

  Administrative Counties    UNIT    VALUE  
0                 Ireland  Number  5149139  
1   Carlow County Council  Number    61968  
2     Dublin City Council  Number   592713

In [433]:
# Show only data that does not have 'Both sexes' in the Sex column
df3 = df3[df3["Sex"] != "Both sexes"]

# And data that does not have 'All ages' in the Single Year of Age column
df3 = df3[df3["Single Year of Age"] != "All ages"]

# Remove Ireland breakdown, and focus solely on the entire counties this time
df3 = df3[df3["Administrative Counties"] != "Ireland"]

# Show first 3 rows to confirm
df3.head(3)

STATISTIC Statistic Label  TLIST(A1)  CensusYear C02199V02655   Sex  \
3297  FY006AC01      Population       2022        2022            1  Male   
3298  FY006AC01      Population       2022        2022            1  Male   
3299  FY006AC01      Population       2022        2022            1  Male   

     C02076V03371 Single Year of Age                          C03789V04537  \
3297          200       Under 1 year  2ae19629-1492-13a3-e055-000000000001   
3298          200       Under 1 year  2ae19629-1433-13a3-e055-000000000001   
3299          200       Under 1 year  2ae19629-149f-13a3-e055-000000000001   

                    Administrative Counties    UNIT  VALUE  
3297                  Carlow County Council  Number    346  
3298                    Dublin City Council  Number   3188  
3299  Dún Laoghaire Rathdown County Council  Number   1269

As with in part 1, the age now also needs to be converted solely to integers.

In [434]:
# Change 'Under 1 year' to '0'
df3["Single Year of Age"] = df3["Single Year of Age"].str.replace('Under 1 year', '0')

# Remove all non-digit characters, leaving only the number
df3["Single Year of Age"] = df3["Single Year of Age"].str.replace('\D', '', regex=True)

# Convert all age values to 64-bit integers
df3['Single Year of Age']=df3['Single Year of Age'].astype('int64')

# Convert all 'Value' numbers to 64-bit integers
df3['VALUE']=df3['VALUE'].astype('int64')

df3.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6262 entries, 3297 to 9791
Data columns (total 12 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   STATISTIC                6262 non-null   object
 1   Statistic Label          6262 non-null   object
 2   TLIST(A1)                6262 non-null   int64 
 3   CensusYear               6262 non-null   int64 
 4   C02199V02655             6262 non-null   object
 5   Sex                      6262 non-null   object
 6   C02076V03371             6262 non-null   object
 7   Single Year of Age       6262 non-null   int64 
 8   C03789V04537             6262 non-null   object
 9   Administrative Counties  6262 non-null   object
 10  UNIT                     6262 non-null   object
 11  VALUE                    6262 non-null   int64 
dtypes: int64(4), object(8)
memory usage: 636.0+ KB


<>:5: SyntaxWarning: invalid escape sequence '\D'
<>:5: SyntaxWarning: invalid escape sequence '\D'
C:\Users\danie\AppData\Local\Temp\ipykernel_19848\3986423257.py:5: SyntaxWarning: invalid escape sequence '\D'
  df3["Single Year of Age"] = df3["Single Year of Age"].str.replace('\D', '', regex=True)


Unrequired columns can now be removed.

In [435]:
# Comfirm all column names
headers = df3.columns.tolist()

# Show list
headers

['STATISTIC',
 'Statistic Label',
 'TLIST(A1)',
 'CensusYear',
 'C02199V02655',
 'Sex',
 'C02076V03371',
 'Single Year of Age',
 'C03789V04537',
 'Administrative Counties',
 'UNIT',
 'VALUE']

In [436]:
# Remove unrequired columns
drop_col_list = ['STATISTIC', 'Statistic Label','TLIST(A1)','CensusYear','C02199V02655','C02076V03371','C03789V04537','UNIT']
df3.drop(columns=drop_col_list, inplace=True)

#print (df.head(3))
df3.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6262 entries, 3297 to 9791
Data columns (total 4 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   Sex                      6262 non-null   object
 1   Single Year of Age       6262 non-null   int64 
 2   Administrative Counties  6262 non-null   object
 3   VALUE                    6262 non-null   int64 
dtypes: int64(2), object(2)
memory usage: 244.6+ KB


In [437]:
# Data can now be converted to pivot table
# Source for multiple columns: https://pandas.pydata.org/docs/user_guide/reshaping.html
df3_anal = pd.pivot_table(df3, 'VALUE',"Single Year of Age",["Sex","Administrative Counties"])

#Show
print (df3_anal.head(3))

# write out the entire file to local machine
# df3_anal.to_csv("population_for_analysis.csv")

Sex                                    Female                       \
Administrative Counties Carlow County Council Cavan County Council   
Single Year of Age                                                   
0                                       353.0                501.0   
1                                       302.0                477.0   
2                                       334.0                520.0   

Sex                                                             \
Administrative Counties Clare County Council Cork City Council   
Single Year of Age                                               
0                                      691.0            1124.0   
1                                      704.0            1136.0   
2                                      744.0            1162.0   

Sex                                                                 \
Administrative Counties Cork County Council Donegal County Council   
Single Year of Age                        

In [438]:
# Specify age
age = 35

# specify range in number of years
range = 5

# Create group within above range
df3_county = df3_anal[(df3_anal.index >= (age-range)) & (df3_anal.index <= (age+range))]

# Show
df3_county

Sex                                    Female                       \
Administrative Counties Carlow County Council Cavan County Council   
Single Year of Age                                                   
30                                      338.0                425.0   
31                                      390.0                457.0   
32                                      384.0                530.0   
33                                      384.0                489.0   
34                                      414.0                513.0   
35                                      446.0                560.0   
36                                      438.0                602.0   
37                                      457.0                604.0   
38                                      483.0                684.0   
39                                      553.0                656.0   
40                                      487.0                630.0   

Sex                                                             \
Administrative Counties Clare County Council Cork City Council   
Single Year of Age                                               
30                                     627.0            1734.0   
31                                     708.0            1814.0   
32                                     688.0            1837.0   
33                                     695.0            1705.0   
34                                     741.0            1774.0   
35                                     787.0            1780.0   
36                                     828.0            1760.0   
37                                     923.0            1872.0   
38                                     955.0            1737.0   
39                                     945.0            1858.0   
40                                     999.0            1879.0   

Sex                                                                 \
Administrative Counties Cork County Council Donegal County Council   
Single Year of Age                                                   
30                                   1873.0                  906.0   
31                                   1970.0                  892.0   
32                                   2050.0                  936.0   
33                                   2023.0                  961.0   
34                                   2216.0                 1077.0   
35                                   2430.0                 1057.0   
36                                   2584.0                 1107.0   
37                                   2611.0                 1115.0   
38                                   2723.0                 1119.0   
39                                   3007.0                 1237.0   
40                                   3058.0                 1293.0   

Sex                                          \
Administrative Counties Dublin City Council   
Single Year of Age                            
30                                   6178.0   
31                                   6079.0   
32                                   5964.0   
33                                   5700.0   
34                                   5545.0   
35                                   5444.0   
36                                   5218.0   
37                                   4902.0   
38                                   5053.0   
39                                   4939.0   
40                                   4809.0   

Sex                                                            \
Administrative Counties Dún Laoghaire Rathdown County Council   
Single Year of Age                                              
30                                                     1601.0   
31                                                     1555.0   
32                                                     1649.0   
33                                                     1581.0   

In [439]:
# Sum county columns (0) and male column (1) together
range3_pop_total = (df3_county.iloc[:].sum())

# Show
range3_pop_total

Sex     Administrative Counties        
Female  Carlow County Council               4774.0
        Cavan County Council                6150.0
        Clare County Council                8896.0
        Cork City Council                  19750.0
        Cork County Council                26545.0
                                            ...   
Male    Tipperary County Council           10882.0
        Waterford City & County Council     8661.0
        Westmeath County Council            7089.0
        Wexford County Council             10824.0
        Wicklow County Council             10338.0
Length: 62, dtype: float64

# End